main.c

```diff
@@ -214,12 +216,30 @@ asm_return_t lz_main(void)
 	/* Set the DEV address for the TB stub to use */
 	bp->tb_dev_map = _u(dev_table);
 
-	sl_stub_entry_offset = _p(bp->code32_start + bp->mle_header + 24);
+	print("\ncode32_start ");
+	print_p(_p(bp->code32_start));
 
-	print("sl_stub_entry_offset:\n");
-	hexdump(sl_stub_entry_offset, 0x100);
+	ki = _p(bp->code32_start + bp->kern_info_offset);
+	mle_header = _p(bp->code32_start + ki->mle_header_offset);
 
-	pm_kernel_entry = _p(bp->code32_start + *sl_stub_entry_offset);
+	if (bp->version             < 0x020f
+	    || ki->header          != KERNEL_INFO_HEADER
+	    || mle_header->uuid[0] != MLE_UUID0
```

Comment:
Sadly, Clang can (and will) miscompile this (Clang devs argue that it is legal under the C spec).  It can logically turn each `||` into `|` to aggressively reduce the number of jumps (/basic blocks).

The most efficient (compact) way to write this is:
```
if (bp->version < 0x020f ||
    (ki = _p(bp->code32_start +
	     bp->kern_info_offset))->header != KERNEL_INFO_HEADER ||
    (mle_header = _p(bp->code32_start +
		     ki->mle_header_offset))->uuid[0] != MLE_UUID0 ||
    ...
```
which moves the assignment of the `ki` and `mle_header` variables until after the sequence point from the preceding `||` which forces the safety check to be first.

However, we should also do some sanity checks that all offsets from `bp->code32_start` fall within the kernel image, or a malformed kernel header can cause us to operate on junk.  Perhaps we want some inline helpers such as `$FOO *get_$FOO_header(...)` which do all the sanity checks, so the outside logic can be:
```
if (bp->version < 0x020f ||
    (ki = get_kern_info_header(bp)) != NULL ||
    (mle_header = get_mle_header(bp, ki)) |= NULL)
    ...
```